olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h

```diff
@@ -25,22 +25,51 @@
 #include <utility>
 
 #include <olp/authentication/AuthenticationCredentials.h>
-#include <olp/authentication/AutoRefreshingToken.h>
+#include <olp/authentication/ErrorResponse.h>
 #include <olp/authentication/Settings.h>
-#include <olp/authentication/TokenEndpoint.h>
-#include <olp/authentication/TokenResult.h>
 #include <olp/authentication/Types.h>
 #include <olp/core/client/CancellationContext.h>
 #include <olp/core/client/OauthToken.h>
 #include <olp/core/http/HttpStatusCode.h>
-#include <olp/core/utils/WarningWorkarounds.h>
 
 namespace olp {
 namespace authentication {
 
-// Needed to avoid endless warnings from TokenRequest/TokenResult
-PORTING_PUSH_WARNINGS()
-PORTING_CLANG_GCC_DISABLE_WARNING("-Wdeprecated-declarations")
+static constexpr auto kDefaultMinimumValidity = 300ll;
+static constexpr auto kDefaultMinimumValiditySeconds =
+    std::chrono::seconds(kDefaultMinimumValidity);
+static constexpr auto kForceRefresh = std::chrono::seconds(0);
+
+template <uint64_t MinimumValidity>
+class TokenProvider;
+
+class AutoRefreshingToken;
+
+namespace internal {
+
+// TokenProviderImpl should only be used as an implementation of TokenProvider
+class TokenProviderImpl {
+  template <uint64_t MinimumValidity>
+  friend class olp::authentication::TokenProvider;
+
+  TokenProviderImpl(Settings settings, std::chrono::seconds minimum_validity);
+  client::OauthTokenResponse operator()(
+      client::CancellationContext& context) const;
+
+  ErrorResponse GetErrorResponse() const;
+
+  int GetHttpStatusCode() const;
+
+  TokenResponse GetResponse(client::CancellationContext& context) const;
+
+  bool IsTokenResponseOK() const;
+
+ private:
+  std::chrono::seconds minimum_validity_{kDefaultMinimumValidity};
+  std::shared_ptr<AutoRefreshingToken> token_;
```

Comment:
AutoRefreshingToken is a pimpl to shared_ptr so once we can hide it lets remove the shared_ptr wrapper.

Reply:
done

Reply:
done